library/modules/EventManager.cpp

```diff
@@ -405,51 +578,92 @@ static void manageJobInitiatedEvent(color_ostream& out) {
         return;
     if (!df::global::job_next_id)
         return;
-    if ( lastJobId == -1 ) {
-        lastJobId = *df::global::job_next_id - 1;
-        return;
+    int32_t tick = df::global::world->frame_counter;
+    int32_t oldest_last_tick = (uint16_t)-1;
+
+    // update new jobs list
+    for ( df::job_list_link* link = df::global::world->jobs.list.next; link != nullptr; link = link->next ) {
+        df::job* job = link->item;
+        // the jobs must come after the last known job id
+        if (job && job->id > lastJobId) {
+            newJobs.emplace(tick, Job::cloneJobStruct(link->item, true));
```

Comment:
Your recent comments about lifecycle management made me look at this a little closer. By cloning the job struct, doesn't this prevent handlers from modifying the job? Or if they do want to modify, do they now need to look the job up themselves? This might hold for any objects that we now make a copy of for lifecycle management purposes.

Reply:
> doesn't this prevent handlers from modifying the job? Or if they do want to modify, do they now need to look the job up themselves?

Yea, if they want to modify the job they would need to look it up. If I go about optimizing all the job* stuff then maybe I can find a way to retain sending the "true" job* when possible for the first two job event types (ie. not completed).

First optimization I think would be to always send the true-job* when the event's tick age is 0. As for the rest.. I think I would need to search for the true-job* from the id.. so ig a whole bunch of binary searches. Maybe I could treat repeating jobs a bit differently and cache them(?) probably safer to just search them up as well.

The completed event type would still need to use cloned-job* though, for [I think] obvious reasons.

> This might hold for any objects that we now make a copy of for lifecycle management purposes.

I haven't taken to managing any other objects. Everything else is purely ids at the root of it. Except the constructions, those structs don't have ids in them (that I could see)

Reply:
Aside from listeners running every tick, and that 0-age edge case.. I think it might be best to just leave the task of finding a modifiable job* to the listener. Though I certainly don't love the idea of sending different sets of pointers to different listeners based on their tick frequency and/or happenstance (0-age).

Reply:
If this were a new API, I might agree, but I think we have to provide the same guarantees as the existing implementation in order to avoid breaking things

Reply:
I shall start by changing the type in the declaration.. let's see. Then 6 hours later we find ourselves mid refactor of all the scanning code.